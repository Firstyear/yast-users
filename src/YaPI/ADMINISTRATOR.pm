package YaPI::ADMINISTRATOR;

use strict;
use YaST::YCP qw(:LOGGING);
use YaPI;

textdomain ("users");

# ------------------- imported modules
YaST::YCP::Import ("MailAliases");
YaST::YCP::Import ("Users");
# -------------------------------------

our $VERSION		= '1.0.0';
our @CAPABILITIES 	= ('SLES11');
our %TYPEINFO;

=item *
C<$hash Read ();>

Returns the information about system administrator (root).
Currently return hash contains the list of mail aliases.

=cut

BEGIN{$TYPEINFO{Read} = ["function",
    [ "map", "string", "any" ]];
}
sub Read {

    my $self	= shift;

    my $root_mail	= MailAliases->GetRootAlias ();
    return {} if !defined $root_mail;

    my @root_aliases	= ();
    foreach my $alias (split (/,/, $root_mail)) {
	$alias	=~ s/[ \t]//g;
	push @root_aliases, $alias;
    }

    return {
	"aliases"	=> \@root_aliases
    }
}

=item *
C<$string Write ($argument_hash);>

write the system adminstrator data. Supported keys of the argument hash are:

    "aliases"	=> list of mail aliases
    "password"	=> new password
    
Returns error message on error.

=cut

BEGIN{$TYPEINFO{Write} = ["function",
    "string",
    [ "map", "string", "any" ]];
}
sub Write {

    my $self	= shift;
    my $args	= shift;
    my $ret	= "";

    if ($args->{"aliases"} && ref ($args->{"aliases"}) eq "ARRAY") {

	my $root_mail	= join (", ", @{$args->{"aliases"}});
	if (!MailAliases->SetRootAlias ($root_mail)) {
	    # error popup
	    $ret = __("An error occurred while setting forwarding for root's mail.");
	    return $ret;
	}
    }
    if ($args->{"password"}) {
	Users->SetRootPassword ($args->{"password"});
	Users->WriteRootPassword ();
    }
    return $ret;
}

