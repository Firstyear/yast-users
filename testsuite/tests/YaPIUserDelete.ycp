/**
 * $Id$
 */

{
    // testedfiles: Users.pm UsersCache.pm USERS.pm

    include "testsuite.ycp";

    import "YaPI::USERS";
    import "Mode";

    map R = $[
	"passwd": $[
	    "local": $[
		"users": $[
		    500: $[
			"uid"		: "hh",
			"uidnumber"	: 500,
			"gidnumber"	: 100,
			"type"		: "local",
		    ],
		    "by_name": $[
			"hh":	500,
		    ],
		    "last_uid": 500,
		    "homes":	$[
			"/home/hh":	1,
		    ],
		    "usernames": $[
			"hh":		1,
		    ],
		    "uids": $[
			500:		1,
		    ]
		],
		"shadow": $[
		    "hh": $[
		    ]
		],
		"groups": $[
		    100: $[
			"cn"	:	"users",
			"gidnumber":	100,
			"type"	:	"local",
			"more_users":	$[
			    "hh":	1,
			]
		    ],
		    1000: $[
			"cn"	:	"gg",
			"gidnumber":	1000,
			"type"	:	"local",
			"userlist":	$[
			    "hh":	1
			]
		    ],
		    1001: $[
			"cn"	:	"gr",
			"gidnumber":	1001,
			"type"	:	"local",
		    ],
		    "by_name": $[
			"users":	100,
			"gg"	:	1000,
			"gr"	:	1001
		    ],
		    "gids": $[
			100:		1,
			1000:		1,
			1001:		1,
		    ],
		    "groupnames": $[
			"users":	1,
			"gg":		1,
			"gr":		1,
		    ]
		],
	    ],
	    "system": $[
		"users": $[
		    0:	$[
			"uid"		: "root",
			"uidnumber"	: 0
		    ],
		    "by_name": $[
			"root"		: 0
		    ],
		    "last_uid": 0,
		    "homes":	$[
		    ],
		    "usernames": $[
			"root"		: 1,
		    ],
		    "uids": $[
			0		: 1
		    ],
		],
		"shadow": $[
		    "root": $[
		    ]
		],
		"groups": $[
		    0: $[
			"cn":	"root",
			"gidnumber":	0,
			"type"	:	"system"
		    ],
		    "by_name": $[
			"root"	: 0
		    ],
		    "gids": $[
			0	: 1
		    ],
		    "groupnames": $[
			"root"	: 1
		    ]
		],
	    ],
	],
	"etc" : $[
	    "fstab": [],
	    "cryptotab": [],
	    "default": $[
		"useradd": $[
		    "home":	"/tmp/local/home",
		    "groups":	"audio,video",
		    "group":	100
		]
	    ],
	],
	"target": $[
	    "stat"	: $[],
	    "size"	: -1,
	    "tmpdir"	: "/tmp/YaST"
	],
	"product": $[
		"features": $[
		    "USE_DESKTOP_SCHEDULER"	: "no",
		    "IO_SCHEDULER"		: "",
		    "ENABLE_AUTOLOGIN"		: "false",
		    "UI_MODE"			: "simple",
		    "EVMS_CONFIG"		: "no"
		]
	],
    ];
    map W = $[];
    map E = $[
	"passwd" : $[
	    "init"	: true
	],
    ];

    DUMP ("==========================================================");

    Mode::SetTest ("test");

    map<string,any> config_map  = $[];

    DUMP ("============ delete non-existent user: ==================");
    config_map["uid"]	= "hhh";
    string error = (string)
	TEST (``(YaPI::USERS::UserDelete (config_map)), [R,W,E], 0);

    DUMP ("============ delete user by name: =======================");

    config_map["uid"]	= "hh";
    TEST (``(YaPI::USERS::UserDelete (config_map)), [R,W,E], 0);

    DUMP ("============ delete user by uid number: ================");

    config_map	= $[
	"uidnumber"	: "500"
    ];
    TEST (``(YaPI::USERS::UserDelete (config_map)), [R,W,E], 0);
}
