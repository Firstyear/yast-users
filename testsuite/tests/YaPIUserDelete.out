Dump	==========================================================
Dump	============ delete non-existent user: ==================
Read	.target.tmpdir "/tmp/YaST"
Read	.etc.default.useradd."expire" 0
Read	.etc.default.useradd."group" 100
Read	.etc.default.useradd."groups" "audio,video"
Read	.etc.default.useradd."home" "/tmp/local/home"
Read	.etc.default.useradd."inactive" 0
Read	.etc.default.useradd."shell" 0
Read	.etc.default.useradd."skel" 0
Read	.target.string "/etc/shells" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.etc.nsswitch_conf.passwd 0
Write	.target.ycp "/var/lib/YaST2/users.ycp" $[] true
Execute	.passwd.init $["base_directory":"/etc", "max_system_gid":"499", "max_system_uid":"499"] true
Read	.passwd.local.users $["by_uidnumber":$[500:$["hh":1]], "hh":$["gidnumber":100, "type":"local", "uid":"hh", "uidnumber":500], "homes":$["/home/hh":1], "last_uid":500, "uids":$[500:1], "usernames":$["hh":1]]
Read	.passwd.local.users.by_uidnumber $[500:$["hh":1]]
Read	.passwd.local.shadow $["hh":$[]]
Read	.passwd.local.groups $["by_gidnumber":$[100:$["users":1], 1000:$["gg":1], 1001:$["gr":1]], "gg":$["cn":"gg", "gidnumber":1000, "type":"local", "userlist":$["hh":1]], "gids":$[100:1, 1000:1, 1001:1], "gr":$["cn":"gr", "gidnumber":1001, "type":"local"], "groupnames":$["gg":1, "gr":1, "users":1], "users":$["cn":"users", "gidnumber":100, "more_users":$["hh":1], "type":"local"]]
Read	.passwd.local.groups.by_gidnumber $[100:$["users":1], 1000:$["gg":1], 1001:$["gr":1]]
Read	.passwd.system.users $["by_uidnumber":$[0:$["root":1]], "homes":$[], "last_uid":0, "root":$["uid":"root", "uidnumber":0], "uids":$[0:1], "usernames":$["root":1]]
Read	.passwd.system.users.by_uidnumber $[0:$["root":1]]
Read	.passwd.system.shadow $["root":$[]]
Read	.passwd.system.groups $["by_gidnumber":$[0:$["root":1]], "gids":$[0:1], "groupnames":$["root":1], "root":$["cn":"root", "gidnumber":0, "type":"system"]]
Read	.passwd.system.groups.by_gidnumber $[0:$["root":1]]
Read	.passwd.passwd.pluslines 0
Read	.passwd.shadow.pluslines 0
Read	.passwd.group.pluslines 0
Read	.passwd.local.users.last_uid 500
Read	.passwd.local.users.homes $["/home/hh":1]
Read	.passwd.local.users.usernames $["hh":1]
Read	.passwd.local.users.uids $[500:1]
Read	.passwd.system.users.last_uid 0
Read	.passwd.system.users.homes $[]
Read	.passwd.system.users.usernames $["root":1]
Read	.passwd.system.users.uids $[0:1]
Read	.passwd.local.groups.gids $[100:1, 1000:1, 1001:1]
Read	.passwd.local.groups.groupnames $["gg":1, "gr":1, "users":1]
Read	.passwd.system.groups.gids $[0:1]
Read	.passwd.system.groups.groupnames $["root":1]
Return	There is no such user.
Dump	============ delete user by name: =======================
Read	.target.tmpdir "/tmp/YaST"
Read	.etc.default.useradd."expire" 0
Read	.etc.default.useradd."group" 100
Read	.etc.default.useradd."groups" "audio,video"
Read	.etc.default.useradd."home" "/tmp/local/home"
Read	.etc.default.useradd."inactive" 0
Read	.etc.default.useradd."shell" 0
Read	.etc.default.useradd."skel" 0
Read	.target.string "/etc/shells" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.etc.nsswitch_conf.passwd 0
Write	.target.ycp "/var/lib/YaST2/users.ycp" $[] true
Execute	.passwd.init $["base_directory":"/etc", "max_system_gid":"499", "max_system_uid":"499"] true
Read	.passwd.local.users $["by_uidnumber":$[500:$["hh":1]], "hh":$["gidnumber":100, "type":"local", "uid":"hh", "uidnumber":500], "homes":$["/home/hh":1], "last_uid":500, "uids":$[500:1], "usernames":$["hh":1]]
Read	.passwd.local.users.by_uidnumber $[500:$["hh":1]]
Read	.passwd.local.shadow $["hh":$[]]
Read	.passwd.local.groups $["by_gidnumber":$[100:$["users":1], 1000:$["gg":1], 1001:$["gr":1]], "gg":$["cn":"gg", "gidnumber":1000, "type":"local", "userlist":$["hh":1]], "gids":$[100:1, 1000:1, 1001:1], "gr":$["cn":"gr", "gidnumber":1001, "type":"local"], "groupnames":$["gg":1, "gr":1, "users":1], "users":$["cn":"users", "gidnumber":100, "more_users":$["hh":1], "type":"local"]]
Read	.passwd.local.groups.by_gidnumber $[100:$["users":1], 1000:$["gg":1], 1001:$["gr":1]]
Read	.passwd.system.users $["by_uidnumber":$[0:$["root":1]], "homes":$[], "last_uid":0, "root":$["uid":"root", "uidnumber":0], "uids":$[0:1], "usernames":$["root":1]]
Read	.passwd.system.users.by_uidnumber $[0:$["root":1]]
Read	.passwd.system.shadow $["root":$[]]
Read	.passwd.system.groups $["by_gidnumber":$[0:$["root":1]], "gids":$[0:1], "groupnames":$["root":1], "root":$["cn":"root", "gidnumber":0, "type":"system"]]
Read	.passwd.system.groups.by_gidnumber $[0:$["root":1]]
Read	.passwd.passwd.pluslines 0
Read	.passwd.shadow.pluslines 0
Read	.passwd.group.pluslines 0
Read	.passwd.local.users.last_uid 500
Read	.passwd.local.users.homes $["/home/hh":1]
Read	.passwd.local.users.usernames $["hh":1]
Read	.passwd.local.users.uids $[500:1]
Read	.passwd.system.users.last_uid 0
Read	.passwd.system.users.homes $[]
Read	.passwd.system.users.usernames $["root":1]
Read	.passwd.system.users.uids $[0:1]
Read	.passwd.local.groups.gids $[100:1, 1000:1, 1001:1]
Read	.passwd.local.groups.groupnames $["gg":1, "gr":1, "users":1]
Read	.passwd.system.groups.gids $[0:1]
Read	.passwd.system.groups.groupnames $["root":1]
Execute	.target.bash "/bin/cp /etc/passwd /etc/passwd.YaST2save" 0
Write	.passwd.users $["local":$["by_uidnumber":$["500":$["hh":"1"]], "homes":$["/home/hh":"1"], "last_uid":"500", "uids":$["500":"1"], "usernames":$["hh":"1"]], "system":$["by_uidnumber":$["0":$["root":"1"]], "homes":$[], "last_uid":"0", "root":$["uid":"root", "uidnumber":"0"], "uids":$["0":"1"], "usernames":$["root":"1"]]] true
Execute	.target.bash "/bin/cp /etc/shadow /etc/shadow.YaST2save" 0
Write	.passwd.shadow $["local":$[], "system":$["root":$[]]] true
Execute	.target.bash "/usr/sbin/nscd -i passwd" 0
Write	.target.ycp "/var/lib/YaST2/users.ycp" $["custom_groups":["local"], "custom_users":["local"], "dont_warn_when_nisserver_notdes":false, "dont_warn_when_uppercase":false] true
Return	
Dump	============ delete user by uid number: ================
Read	.target.tmpdir "/tmp/YaST"
Read	.etc.default.useradd."expire" 0
Read	.etc.default.useradd."group" 100
Read	.etc.default.useradd."groups" "audio,video"
Read	.etc.default.useradd."home" "/tmp/local/home"
Read	.etc.default.useradd."inactive" 0
Read	.etc.default.useradd."shell" 0
Read	.etc.default.useradd."skel" 0
Read	.target.string "/etc/shells" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.etc.nsswitch_conf.passwd 0
Write	.target.ycp "/var/lib/YaST2/users.ycp" $[] true
Execute	.passwd.init $["base_directory":"/etc", "max_system_gid":"499", "max_system_uid":"499"] true
Read	.passwd.local.users $["by_uidnumber":$[500:$["hh":1]], "hh":$["gidnumber":100, "type":"local", "uid":"hh", "uidnumber":500], "homes":$["/home/hh":1], "last_uid":500, "uids":$[500:1], "usernames":$["hh":1]]
Read	.passwd.local.users.by_uidnumber $[500:$["hh":1]]
Read	.passwd.local.shadow $["hh":$[]]
Read	.passwd.local.groups $["by_gidnumber":$[100:$["users":1], 1000:$["gg":1], 1001:$["gr":1]], "gg":$["cn":"gg", "gidnumber":1000, "type":"local", "userlist":$["hh":1]], "gids":$[100:1, 1000:1, 1001:1], "gr":$["cn":"gr", "gidnumber":1001, "type":"local"], "groupnames":$["gg":1, "gr":1, "users":1], "users":$["cn":"users", "gidnumber":100, "more_users":$["hh":1], "type":"local"]]
Read	.passwd.local.groups.by_gidnumber $[100:$["users":1], 1000:$["gg":1], 1001:$["gr":1]]
Read	.passwd.system.users $["by_uidnumber":$[0:$["root":1]], "homes":$[], "last_uid":0, "root":$["uid":"root", "uidnumber":0], "uids":$[0:1], "usernames":$["root":1]]
Read	.passwd.system.users.by_uidnumber $[0:$["root":1]]
Read	.passwd.system.shadow $["root":$[]]
Read	.passwd.system.groups $["by_gidnumber":$[0:$["root":1]], "gids":$[0:1], "groupnames":$["root":1], "root":$["cn":"root", "gidnumber":0, "type":"system"]]
Read	.passwd.system.groups.by_gidnumber $[0:$["root":1]]
Read	.passwd.passwd.pluslines 0
Read	.passwd.shadow.pluslines 0
Read	.passwd.group.pluslines 0
Read	.passwd.local.users.last_uid 500
Read	.passwd.local.users.homes $["/home/hh":1]
Read	.passwd.local.users.usernames $["hh":1]
Read	.passwd.local.users.uids $[500:1]
Read	.passwd.system.users.last_uid 0
Read	.passwd.system.users.homes $[]
Read	.passwd.system.users.usernames $["root":1]
Read	.passwd.system.users.uids $[0:1]
Read	.passwd.local.groups.gids $[100:1, 1000:1, 1001:1]
Read	.passwd.local.groups.groupnames $["gg":1, "gr":1, "users":1]
Read	.passwd.system.groups.gids $[0:1]
Read	.passwd.system.groups.groupnames $["root":1]
Execute	.target.bash "/bin/cp /etc/passwd /etc/passwd.YaST2save" 0
Write	.passwd.users $["local":$["by_uidnumber":$["500":$["hh":"1"]], "homes":$["/home/hh":"1"], "last_uid":"500", "uids":$["500":"1"], "usernames":$["hh":"1"]], "system":$["by_uidnumber":$["0":$["root":"1"]], "homes":$[], "last_uid":"0", "root":$["uid":"root", "uidnumber":"0"], "uids":$["0":"1"], "usernames":$["root":"1"]]] true
Execute	.target.bash "/bin/cp /etc/shadow /etc/shadow.YaST2save" 0
Write	.passwd.shadow $["local":$[], "system":$["root":$[]]] true
Execute	.target.bash "/usr/sbin/nscd -i passwd" 0
Write	.target.ycp "/var/lib/YaST2/users.ycp" $["custom_groups":["local"], "custom_users":["local"], "dont_warn_when_nisserver_notdes":false, "dont_warn_when_uppercase":false] true
Return	
